keps/sig-api-machinery/4355-coordinated-leader-election/README.md

```diff
@@ -0,0 +1,1190 @@
+  
+
+### Non-Goals
+
+- Change the default leader election for components.
+
+## Proposal
+
+- Offer an opt-in leader election mechanism to:
+  - Elect the candidate with the oldest version available.
+  - Provide a way to preempt the current leader.
+  - Reuse the existing lease mechanism as much as possible.
+
+### Component Identity Leases
+
+Components will create identity leases similar to those used by apiserver
+identity, e.g.:
+
+```yaml
+apiVersion: coordination.k8s.io/v1
+kind: Lease
+metadata:
+  annotations:
+    coordination.k8s.io/binary-version: "1.29"
+    coordination.k8s.io/can-lead-leases: kube-system/some-custom-controller
+    coordination.k8s.io/compatibility-version: "1.29"
+  name: some-custom-controller-0001A
+  namespace: kube-system
+spec:
+  holderIdentity: some-custom-controller-0001A
+  leaseDurationSeconds: 10
+  renewTime: "2023-12-05T02:33:08.685777Z"
+```
+
+A component identity lease announces candidacy for leadership by including
+`coordination.k8s.io/can-lead-leases` it is identity lease. If the lease
+expires, the component is considered unavailable for leader election purposes.
+
+### Coordinated Election Controller
+
+A new Coordinated Election Controller will reconcile component Leader Leases
+(primary resource) and Identity Leases (secondary resource, changes trigger
+reconciliation of related leader leases).
+
+Coordinated Election Controller reconciliation loop:
+
+- If no leader lease exists for a components:
+  - Elect leader from candidates by preparing a freshly renewed lease with:
+    - `spec.holderIdentity` set to the identity of the elected leader
+    - `coordination.k8s.io/elected-by: leader-election-controller` (to make
+      lease types easy to disambiguate)
+- If there is a better candidate than current leader:
+  - Sets `coordination.k8s.io/end-of-term: true` on the leader lease, signaling
+    that the leader should stop renewing the lease and yield leadership
+
+```mermaid
+flowchart TD
+   A[Reconcile] --> |Process Leader Lease| B
+   B{Lease Status?} --> |Better Leader Exists| D
+   B --> |Expired/Missing| E
+   D[End Lease Term]
+   E[Elect Leader]
+```
+
+Example of a lease created by Coordinated Election Controller:
+
+```yaml
+apiVersion: coordination.k8s.io/v1
+kind: Lease
+metadata:
+  annotations:
+    coordination.k8s.io/elected-by: coordinated-election-controller
+  name: some-custom-controller
+  namespace: kube-system
+spec:
+  holderIdentity: controller-a
+  leaseDurationSeconds: 10
+  leaseTransitions: 0
+  renewTime: "2023-12-05T18:58:31.295467Z"
+```
+
+The Coordinated Election Controller will run in the kube-apiserver.
+
+In an HA configuration, the Coordinated Leader Election Controller will have its
+own lease similar to how other leader elected controllers behaves today. It will
+be responsible for renewing its own lease and gracefully shutdown if the lease
+is expired. Only one instance of the coordinated leader election controller will
```

Comment:
does this matter? If the logic is deterministic and we make use of optimistic concurrency cleverly, we wouldn't need any coordination of the election coordinators, would we?